AssertionError after several variant QC steps.

To report a bug, fill in the information below. 
For support and feature requests, please use the discussion forum: http://discuss.hail.is/

-------------------------------------------------------------------------------------------

### Hail version:
devel-6bb4670

### What you did:
A number of variant QC steps, then a `vds.write`
The error is probably caused by one of the previous steps. If it helps I can comment out earlier parts to narrow down what actually triggers the error.

### What went wrong (all error messages here, including the full java stack trace):
```
[Stage 6:>                                                       (0 + 8) / 5000]
[Stage 6:>                                                       (0 + 4) / 5000]
[Stage 6:>                                                       (0 + 8) / 5000]Traceback (most recent call last):
  File "/home/hail/hail.zip/hail/utils/java.py", line 185, in handle_py4j
  File "/home/hail/hail.zip/hail/table.py", line 1058, in aggregate
  File "/usr/lib/spark/python/lib/py4j-0.10.4-src.zip/py4j/java_gateway.py", line 1133, in __call__
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/sql/utils.py", line 63, in deco
  File "/usr/lib/spark/python/lib/py4j-0.10.4-src.zip/py4j/protocol.py", line 319, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling o30335.query.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 7 in stage 6.0 failed 20 times, most recent failure: Lost task 7.19 in stage 6.0 (TID 179, robert1-w-0.c.ccdg-wgs.internal, executor 4): java.lang.AssertionError: assertion failed
                    at scala.Predef$.assert(Predef.scala:156)
                    at is.hail.annotations.RegionValueBuilder.endStruct(RegionValueBuilder.scala:109)
                    at is.hail.variant.MatrixTable$$anonfun$filterGenotypes$1$$anonfun$apply$80.apply(MatrixTable.scala:2645)
                    at is.hail.variant.MatrixTable$$anonfun$filterGenotypes$1$$anonfun$apply$80.apply(MatrixTable.scala:2615)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at scala.collection.Iterator$class.foreach(Iterator.scala:893)
                    at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
                    at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:157)
                    at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1336)
                    at scala.collection.TraversableOnce$class.aggregate(TraversableOnce.scala:214)
                    at scala.collection.AbstractIterator.aggregate(Iterator.scala:1336)
                    at org.apache.spark.rdd.RDD$$anonfun$aggregate$1$$anonfun$22.apply(RDD.scala:1113)
                    at org.apache.spark.rdd.RDD$$anonfun$aggregate$1$$anonfun$22.apply(RDD.scala:1113)
                    at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:2118)
                    at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:2118)
                    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
                    at org.apache.spark.scheduler.Task.run(Task.scala:108)
                    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
                    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
                    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
                    at java.lang.Thread.run(Thread.java:748)

Driver stacktrace:
                    at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1499)
                    at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1487)
                    at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1486)
                    at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
                    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
                    at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1486)
                    at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
                    at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
                    at scala.Option.foreach(Option.scala:257)
                    at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:814)
                    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1714)
                    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1669)
                    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1658)
                    at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
                    at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:630)
                    at org.apache.spark.SparkContext.runJob(SparkContext.scala:2022)
                    at org.apache.spark.SparkContext.runJob(SparkContext.scala:2119)
                    at org.apache.spark.rdd.RDD$$anonfun$aggregate$1.apply(RDD.scala:1115)
                    at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
                    at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
                    at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
                    at org.apache.spark.rdd.RDD.aggregate(RDD.scala:1108)
                    at is.hail.table.Table.query(Table.scala:339)
                    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
                    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                    at java.lang.reflect.Method.invoke(Method.java:498)
                    at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
                    at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
                    at py4j.Gateway.invoke(Gateway.java:280)
                    at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
                    at py4j.commands.CallCommand.execute(CallCommand.java:79)
                    at py4j.GatewayConnection.run(GatewayConnection.java:214)
                    at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.AssertionError: assertion failed
                    at scala.Predef$.assert(Predef.scala:156)
                    at is.hail.annotations.RegionValueBuilder.endStruct(RegionValueBuilder.scala:109)
                    at is.hail.variant.MatrixTable$$anonfun$filterGenotypes$1$$anonfun$apply$80.apply(MatrixTable.scala:2645)
                    at is.hail.variant.MatrixTable$$anonfun$filterGenotypes$1$$anonfun$apply$80.apply(MatrixTable.scala:2615)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at scala.collection.Iterator$class.foreach(Iterator.scala:893)
                    at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
                    at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:157)
                    at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1336)
                    at scala.collection.TraversableOnce$class.aggregate(TraversableOnce.scala:214)
                    at scala.collection.AbstractIterator.aggregate(Iterator.scala:1336)
                    at org.apache.spark.rdd.RDD$$anonfun$aggregate$1$$anonfun$22.apply(RDD.scala:1113)
                    at org.apache.spark.rdd.RDD$$anonfun$aggregate$1$$anonfun$22.apply(RDD.scala:1113)
                    at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:2118)
                    at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:2118)
                    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
                    at org.apache.spark.scheduler.Task.run(Task.scala:108)
                    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
                    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
                    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
                    ... 1 more


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/b09ec92a-49f4-4d16-ad6d-efc5a5805e92/05_variant_qc.py", line 201, in <module>
    cumcounts = {'step0': rt.aggregate(hl.agg.sum(hl.cond(rt.qccum.step0, 1, 0))),
  File "<decorator-gen-519>", line 2, in aggregate
  File "/home/hail/hail.zip/hail/utils/java.py", line 191, in handle_py4j
hail.utils.java.FatalError: AssertionError: assertion failed

Java stack trace:
org.apache.spark.SparkException: Job aborted due to stage failure: Task 7 in stage 6.0 failed 20 times, most recent failure: Lost task 7.19 in stage 6.0 (TID 179, robert1-w-0.c.ccdg-wgs.internal, executor 4): java.lang.AssertionError: assertion failed
                    at scala.Predef$.assert(Predef.scala:156)
                    at is.hail.annotations.RegionValueBuilder.endStruct(RegionValueBuilder.scala:109)
                    at is.hail.variant.MatrixTable$$anonfun$filterGenotypes$1$$anonfun$apply$80.apply(MatrixTable.scala:2645)
                    at is.hail.variant.MatrixTable$$anonfun$filterGenotypes$1$$anonfun$apply$80.apply(MatrixTable.scala:2615)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at scala.collection.Iterator$class.foreach(Iterator.scala:893)
                    at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
                    at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:157)
                    at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1336)
                    at scala.collection.TraversableOnce$class.aggregate(TraversableOnce.scala:214)
                    at scala.collection.AbstractIterator.aggregate(Iterator.scala:1336)
                    at org.apache.spark.rdd.RDD$$anonfun$aggregate$1$$anonfun$22.apply(RDD.scala:1113)
                    at org.apache.spark.rdd.RDD$$anonfun$aggregate$1$$anonfun$22.apply(RDD.scala:1113)
                    at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:2118)
                    at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:2118)
                    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
                    at org.apache.spark.scheduler.Task.run(Task.scala:108)
                    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
                    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
                    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
                    at java.lang.Thread.run(Thread.java:748)

Driver stacktrace:
                    at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1499)
                    at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1487)
                    at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1486)
                    at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
                    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
                    at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1486)
                    at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
                    at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
                    at scala.Option.foreach(Option.scala:257)
                    at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:814)
                    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1714)
                    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1669)
                    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1658)
                    at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
                    at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:630)
                    at org.apache.spark.SparkContext.runJob(SparkContext.scala:2022)
                    at org.apache.spark.SparkContext.runJob(SparkContext.scala:2119)
                    at org.apache.spark.rdd.RDD$$anonfun$aggregate$1.apply(RDD.scala:1115)
                    at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
                    at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
                    at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
                    at org.apache.spark.rdd.RDD.aggregate(RDD.scala:1108)
                    at is.hail.table.Table.query(Table.scala:339)
                    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
                    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                    at java.lang.reflect.Method.invoke(Method.java:498)
                    at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
                    at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
                    at py4j.Gateway.invoke(Gateway.java:280)
                    at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
                    at py4j.commands.CallCommand.execute(CallCommand.java:79)
                    at py4j.GatewayConnection.run(GatewayConnection.java:214)
                    at java.lang.Thread.run(Thread.java:748)java.lang.AssertionError: assertion failed
                    at scala.Predef$.assert(Predef.scala:156)
                    at is.hail.annotations.RegionValueBuilder.endStruct(RegionValueBuilder.scala:109)
                    at is.hail.variant.MatrixTable$$anonfun$filterGenotypes$1$$anonfun$apply$80.apply(MatrixTable.scala:2645)
                    at is.hail.variant.MatrixTable$$anonfun$filterGenotypes$1$$anonfun$apply$80.apply(MatrixTable.scala:2615)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:736)
                    at is.hail.rvd.OrderedRVD$$anonfun$apply$6$$anon$5.next(OrderedRVD.scala:730)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
                    at scala.collection.Iterator$class.foreach(Iterator.scala:893)
                    at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
                    at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:157)
                    at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1336)
                    at scala.collection.TraversableOnce$class.aggregate(TraversableOnce.scala:214)
                    at scala.collection.AbstractIterator.aggregate(Iterator.scala:1336)
                    at org.apache.spark.rdd.RDD$$anonfun$aggregate$1$$anonfun$22.apply(RDD.scala:1113)
                    at org.apache.spark.rdd.RDD$$anonfun$aggregate$1$$anonfun$22.apply(RDD.scala:1113)
                    at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:2118)
                    at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:2118)
                    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
                    at org.apache.spark.scheduler.Task.run(Task.scala:108)
                    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
                    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
                    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
                    at java.lang.Thread.run(Thread.java:748)


Hail version: devel-6bb4670
Error summary: AssertionError: assertion failed
ERROR: (gcloud.dataproc.jobs.submit.pyspark) Job [b09ec92a-49f4-4d16-ad6d-efc5a5805e92] entered state [ERROR] while waiting for [DONE].
Submitting to cluster 'robert1'...
gcloud command:
gcloud dataproc jobs submit pyspark hail2/05_variant_qc.py \
    --cluster=robert1 \
    --files= \
    --properties= \
    -- \
    onep
```

--- a/python/hail/tests/test_api.py
+++ b/python/hail/tests/test_api.py
@@ -806,6 +806,15 @@ class MatrixTests(unittest.TestCase):
         et = ds.annotate_entries(entry_idx = 10 * ds.row_idx + ds.col_idx).entries().add_index()
         self.assertTrue(et.all(et.idx == et.entry_idx))
 
+    def test_filter_entries(self):
+        ds = hl.utils.range_matrix_table(100, 10)
+        ds = ds.annotate_rows(foo = 5) # triggered a RV bug
+        ds = ds.annotate_cols(bar = 5)
+        ds = ds.filter_entries((ds.col_idx * ds.row_idx) % 4 == 0)
+
+        entries = ds.entries()
+        self.assertTrue(entries.all((entries.col_idx * entries.row_idx) % 4 == 0))
+
 
 class GroupedMatrixTests(unittest.TestCase):
 

--- a/src/main/scala/is/hail/variant/MatrixTable.scala
+++ b/src/main/scala/is/hail/variant/MatrixTable.scala
@@ -1684,7 +1684,7 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
     newGlobals: Annotation = globals)(newEntryType: TStruct,
     inserter: (PC, RegionValue, RegionValueBuilder) => Unit): MatrixTable = {
     insertIntoRow(makePartitionContext, newColType, newColKey, newColValues, newGlobalType, newGlobals)(
-      TArray(newEntryType), MatrixType.entriesIdentifier, inserter)
+      TArray(newEntryType, required = true), MatrixType.entriesIdentifier, inserter)
   }
 
   def insertIntoRow[PC](makePartitionContext: () => PC, newColType: TStruct = colType,
@@ -2607,53 +2607,48 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
     val f: () => java.lang.Boolean = Parser.parseTypedExpr[java.lang.Boolean](filterExpr, ec)
 
     val localKeep = keep
-    val localRVRowType = rvRowType
+    val fullRowType = rvRowType
     val localNSamples = numCols
     val localEntryType = entryType
     val localColValuesBc = colValuesBc
     val localEntriesIndex = entriesIndex
+    val localEntriesType = matrixType.entryArrayType
 
-    copy2(
-      rvd = rvd.mapPartitionsPreservesPartitioning(rvd.typ) { it =>
-        val rvb = new RegionValueBuilder()
-        val rv2 = RegionValue()
-        val fullRow = new UnsafeRow(localRVRowType)
+    insertEntries(() => {
+      val fullRow = new UnsafeRow(fullRowType)
+      val row = fullRow.deleteField(localEntriesIndex)
+      (fullRow, row)
+    })(localEntryType.copy(required = false), { case ((fullRow, row), rv, rvb) =>
+      fullRow.set(rv)
+      val entries = fullRow.getAs[IndexedSeq[Annotation]](localEntriesIndex)
+      val entriesOffset = fullRowType.loadField(rv, localEntriesIndex)
 
-        it.map { rv =>
-          fullRow.set(rv)
-          val row = fullRow.deleteField(localEntriesIndex)
+      rvb.startArray(localNSamples)
 
-          rvb.set(rv.region)
-          rvb.start(localRVRowType)
-          rvb.startStruct()
-
-          var i = 0
-          while (i < localEntriesIndex) {
-            rvb.addField(localRVRowType, rv, i)
-            i += 1
+      var i = 0
+      while (i < localNSamples) {
+        val entry = entries(i)
+        ec.setAll(row,
+          localColValuesBc.value(i),
+          entry)
+        if (Filter.boxedKeepThis(f(), localKeep)) {
+          val isDefined = localEntriesType.isElementDefined(rv.region, entriesOffset, i)
+          if (!isDefined)
+            rvb.setMissing()
+          else {
+            // can't use addElement because we could be losing requiredness
+            val elementOffset = localEntriesType.loadElement(rv.region, entriesOffset, i)
+            rvb.startStruct()
+            rvb.addAllFields(localEntryType, rv.region, elementOffset)
+            rvb.endStruct()
           }
+        } else
+          rvb.setMissing()
 
-          val gs = fullRow.getAs[IndexedSeq[Any]](localEntriesIndex)
-
-          rvb.startArray(localNSamples)
-          i = 0
-          while (i < localNSamples) {
-            val sa = localColValuesBc.value(i)
-            val g = gs(i)
-            ec.setAll(row, sa, g)
-            if (Filter.boxedKeepThis(f(), localKeep))
-              rvb.addAnnotation(localEntryType, g)
-            else
-              rvb.setMissing()
-
-            i += 1
-          }
-          rvb.endArray()
-          rvb.endStruct()
-          rv2.set(rv.region, rvb.end())
-          rv2
-        }
-      })
+        i += 1
+      }
+      rvb.endArray()
+    })
   }
 
 

--- a/src/main/scala/is/hail/variant/MatrixTable.scala
+++ b/src/main/scala/is/hail/variant/MatrixTable.scala
@@ -1684,7 +1684,7 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
     newGlobals: Annotation = globals)(newEntryType: TStruct,
     inserter: (PC, RegionValue, RegionValueBuilder) => Unit): MatrixTable = {
     insertIntoRow(makePartitionContext, newColType, newColKey, newColValues, newGlobalType, newGlobals)(
-      TArray(newEntryType, required = true), MatrixType.entriesIdentifier, inserter)
+      TArray(newEntryType), MatrixType.entriesIdentifier, inserter)
   }
 
   def insertIntoRow[PC](makePartitionContext: () => PC, newColType: TStruct = colType,

